### DISCUSSION TITLE
Request::createFromGlobals doesn't define session

### DISCUSSION THREAD

**Discussion by indapublic** [symfony/symfony#45375] [Q&A] [answered]
```
public function __invoke(Request $request1): Response
{
	$request2 = Request::createFromGlobals();

	dd(
		$request1->hasSession(),
		$request2->hasSession(),
		$request1->getSession(),
		$request2->getSession()
	);
}
```

`$request1` has session and `$request1->getSession()` will return `Session` object.

`$request2` hasn't session and `$request2->getSession()` will throw exception `Session has not been set.`

Because no session is defined [here](https://github.com/symfony/http-foundation/blob/5.4/Request.php#L276).

As answered [here](https://github.com/symfony/symfony/issues/45369#issuecomment-1033619301), it's expected behaviour. Can I ask why?

**Comment by mbabker** [comment, marked as answer]
`Request::createFromGlobals()` creates a new `Request` object using data from most of the superglobals (`$_GET`, `$_POST`, `$_COOKIE`, `$_FILES`, and `$_SERVER` specifically).  Creating a `Request` instance on its own does not add in session management.  The HttpKernel component takes care of this with the `SessionListener` event listener for the `kernel.request` event.

**Comment by indapublic** [comment]
Thanks @mbabker Yes, i understood how it's working. I just wondering WHY it's working it that way.
I'f I'm passing my session cookie - `createFromGlobals` doesn't linked any session (but dependency injection for example does. Why this difference?)

So my question was like not "Why doesn't work here" but "Why Symfony realised that behaviour".

Thanks

**Reply by mbabker** [reply]
`Request::createFromGlobals()` on its own does not arbitrarily add a session to the request object.  The `Request` object requires manually adding the session via `setSession()` or `setSessionFactory()`.  The [`Symfony\Component\HttpKernel\EventListener\SessionListener`](https://github.com/symfony/symfony/blob/6.0/src/Symfony/Component/HttpKernel/EventListener/SessionListener.php) class takes the session info from your configuration and injects it into the request during the `kernel.request` event.  So, when a request is created, and even very early in the request-response cycle, the request does not have a session attached to it; it is only added to the request in one of the first event listeners from the HTTP kernel after the environment is booted up (i.e. container created), and this is so the session can be built using your configuration.

If it didn't work this way, you'd have to create the session yourself in your `public/index.php` file, and you'd have to do it without any form of dependency injection or ability to use any of the configuration info.
